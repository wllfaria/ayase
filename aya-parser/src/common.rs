use nom::branch::alt;
use nom::bytes::complete::{tag, tag_no_case};
use nom::character::complete::{alpha1, alphanumeric1, anychar, char, hex_digit1, multispace0, one_of, space0};
use nom::combinator::{map, opt, peek as nom_peek, recognize};
use nom::multi::{many0_count, separated_list1};
use nom::sequence::{delimited, pair, terminated};
use nom::IResult;

use crate::types::{Ast, Operator};

pub mod precedences {
    pub const BASE: u8 = 0;
    pub const SUM: u8 = 1;
    pub const MUL: u8 = 2;
}

pub fn get_precedence(operator: Operator) -> u8 {
    match operator {
        Operator::Add | Operator::Sub => precedences::SUM,
        Operator::Mul => precedences::MUL,
    }
}

pub fn ws<'parser, O, F: Fn(&'parser str) -> IResult<&'parser str, O>>(
    inner: F,
) -> impl FnMut(&'parser str) -> IResult<&'parser str, O> {
    delimited(space0, inner, space0)
}

pub fn is_terminator(ch: char) -> bool {
    matches!(ch, ']' | ')')
}

pub fn peek(input: &str) -> IResult<&str, char> {
    nom_peek(anychar)(input)
}

pub fn general_register(input: &str) -> IResult<&str, &str> {
    recognize(pair(tag_no_case("r"), one_of("12345678")))(input)
}

pub fn register(input: &str) -> IResult<&str, Ast> {
    map(
        alt((
            general_register,
            tag_no_case("ip"),
            tag_no_case("sp"),
            tag_no_case("fp"),
            tag_no_case("ret"),
        )),
        Ast::Register,
    )(input)
}

pub fn identifier(input: &str) -> IResult<&str, &str> {
    recognize(pair(
        alt((alpha1, tag("_"))),
        many0_count(alt((alphanumeric1, tag("_")))),
    ))(input)
}

pub fn variable(input: &str) -> IResult<&str, Ast> {
    let (input, _) = tag("!")(input)?;
    map(identifier, Ast::Var)(input)
}

pub fn label(input: &str) -> IResult<&str, Ast> {
    let (input, _) = multispace0(input)?;
    let (input, label) = map(terminated(identifier, char(':')), Ast::Label)(input)?;
    let (input, _) = multispace0(input)?;
    Ok((input, label))
}

pub fn constant(input: &str) -> IResult<&str, Ast> {
    let (input, _) = multispace0(input)?;
    let (input, exported) = opt(char('+'))(input)?;
    let exported = exported.is_some();
    let (input, _) = tag("const")(input)?;
    let (input, _) = space0(input)?;

    let (input, name) = identifier(input)?;
    let (input, _) = space0(input)?;
    let (input, _) = char('=')(input)?;
    let (input, _) = space0(input)?;
    let (input, val) = hex_literal(input)?;
    let (input, _) = multispace0(input)?;

    Ok((
        input,
        Ast::Const {
            name,
            exported,
            value: Box::new(val),
        },
    ))
}

pub fn data(input: &str) -> IResult<&str, Ast> {
    let (input, _) = multispace0(input)?;
    let (input, exported) = opt(char('+'))(input)?;
    let exported = exported.is_some();

    let (input, _) = tag("data")(input)?;
    let (input, size) = map(alt((tag("8"), tag("16"))), |s: &str| s.parse::<u8>().unwrap())(input)?;

    let (input, _) = space0(input)?;
    let (input, name) = identifier(input)?;
    let (input, _) = space0(input)?;
    let (input, _) = char('=')(input)?;
    let (input, _) = space0(input)?;
    let (input, _) = char('{')(input)?;
    let (input, _) = space0(input)?;

    let (input, values) = terminated(
        separated_list1(delimited(space0, tag(","), space0), hex_literal),
        opt(tag(",")),
    )(input)?;

    let (input, _) = space0(input)?;
    let (input, _) = char('}')(input)?;
    let (input, _) = multispace0(input)?;

    Ok((
        input,
        Ast::Data {
            name,
            size,
            exported,
            values,
        },
    ))
}

pub fn hex_literal(input: &str) -> IResult<&str, Ast> {
    let (input, _) = tag("$")(input)?;
    map(hex_digit1, Ast::HexLiteral)(input)
}

pub fn address(input: &str) -> IResult<&str, Ast> {
    let (input, _) = tag("&")(input)?;
    map(hex_digit1, Ast::Address)(input)
}

pub fn operator(input: &str) -> IResult<&str, Ast> {
    map(
        alt((
            map(char('+'), |_| Operator::Add),
            map(char('-'), |_| Operator::Sub),
            map(char('*'), |_| Operator::Mul),
        )),
        Ast::Operator,
    )(input)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_exported_data_8() {
        let input = "+data8 some_name = { $01,$02  , $03   , $04 }";
        let (_, result) = data(input).unwrap();
        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_private_data_8() {
        let input = "data8 some_name = { $01,$02  , $03   , $04 }";
        let (_, result) = data(input).unwrap();
        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_exported_data_16() {
        let input = "+data16 some_name = { $0102  , $0304 }";
        let (_, result) = data(input).unwrap();
        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_private_data_16() {
        let input = "data16 some_name = { $0102  , $0304 }";
        let (_, result) = data(input).unwrap();
        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_exported_const() {
        let input = "+const some_name = $0102";
        let (_, result) = constant(input).unwrap();
        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_private_const() {
        let input = "const some_name = $0102";
        let (_, result) = constant(input).unwrap();
        insta::assert_debug_snapshot!(result);
    }
}
