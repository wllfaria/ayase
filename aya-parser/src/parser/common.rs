use std::ops::Range;

use super::error::{bail, unexpected_eof, unexpected_token};
use super::Result;
use crate::ast::ByteOffset;
use crate::lexer::{Kind, Lexer, Token, TransposeRef};

pub fn expect<S: AsRef<str>>(
    expected: Kind,
    lexer: &mut Lexer,
    source: S,
    help: S,
    message: S,
) -> miette::Result<ByteOffset> {
    let current = lexer.next().transpose()?;
    let Some(current) = current else {
        unreachable!();
    };

    if current.kind != expected {
        return Err(bail(source.as_ref(), help.as_ref(), message.as_ref(), current.offset()));
    }

    Ok(current.offset())
}

pub fn expect_fail<S: AsRef<str>>(expected: Kind, lexer: &mut Lexer, source: S) -> miette::Result<bool> {
    let Ok(Some(next)) = lexer.peek().transpose() else {
        let Err(err) = lexer.next().transpose() else {
            return unexpected_eof(source.as_ref(), "unterminated import statement");
        };
        return Err(err);
    };

    if next.kind != expected {
        return unexpected_token(source.as_ref(), next);
    }

    lexer.next().transpose()?;
    Ok(false)
}

pub fn parse_identifier<S: AsRef<str>>(source: S, lexer: &mut Lexer, help: S, message: S) -> Result<ByteOffset> {
    expect(Kind::Ident, lexer, source.as_ref(), help.as_ref(), message.as_ref())
}

pub fn parse_register<S: AsRef<str>>(source: S, lexer: &mut Lexer) -> Result<ByteOffset> {
    let offset = parse_identifier(source.as_ref(), lexer, "", "")?;
    let name = &source.as_ref()[Range::<usize>::from(offset)];
    match name {
        "acc" | "ip" | "r1" | "r2" | "r3" | "r4" | "r5" | "r6" | "r7" | "r8" | "sp" | "fp" => Ok(offset),
        _ => unexpected_token(source.as_ref(), &Token::from_ident(name, offset.start, offset.end)),
    }
}

pub fn parse_hex_lit<S: AsRef<str>>(source: S, lexer: &mut Lexer, help: S, message: S) -> Result<ByteOffset> {
    expect(Kind::HexNumber, lexer, source.as_ref(), help.as_ref(), message.as_ref())
}

pub fn parse_string<S: AsRef<str>>(source: S, lexer: &mut Lexer, help: S, message: S) -> Result<ByteOffset> {
    expect(Kind::String, lexer, source.as_ref(), help.as_ref(), message.as_ref())
}

pub fn parse_variable<S: AsRef<str>>(source: S, lexer: &mut Lexer, help: S, message: S) -> Result<ByteOffset> {
    expect(Kind::Bang, lexer, source.as_ref(), help.as_ref(), message.as_ref())?;
    expect(Kind::Ident, lexer, source.as_ref(), help.as_ref(), message.as_ref())
}

pub fn parse_keyword<S: AsRef<str>>(source: S, lexer: &mut Lexer, expected: Kind) -> Result<ByteOffset> {
    let next = lexer.next().transpose()?;
    let Some(next) = next else {
        return unexpected_eof(source.as_ref(), "invalid eof");
    };

    if next.kind != expected {
        return unexpected_token(source.as_ref(), &next);
    }

    Ok(next.offset())
}
