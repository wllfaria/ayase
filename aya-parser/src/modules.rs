use nom::branch::alt;
use nom::bytes::complete::tag;
use nom::character::complete::{alphanumeric1, multispace0, space0};
use nom::combinator::{map, recognize};
use nom::multi::{many0, many1};
use nom::sequence::{delimited, separated_pair, terminated, tuple};
use nom::IResult;

use crate::common::{address, hex_literal, identifier, variable, ws};
use crate::Ast;

fn parse_path(input: &str) -> IResult<&str, &str> {
    delimited(
        tag("\""),
        recognize(many1(alt((alphanumeric1, tag("/"), tag("."), tag("_"))))),
        tag("\""),
    )(input)
}

fn field_accessor(input: &str) -> IResult<&str, Ast> {
    let (input, module) = identifier(input)?;
    let (input, _) = tag(".")(input)?;
    let (input, field) = identifier(input)?;

    Ok((input, Ast::FieldAccessor { module, field }))
}

fn bracketed_value(input: &str) -> IResult<&str, Ast> {
    let (input, _) = tag("[")(input)?;
    let (input, _) = space0(input)?;

    let (input, value) = alt((field_accessor, variable))(input)?;

    let (input, _) = tag("]")(input)?;
    let (input, _) = space0(input)?;

    Ok((input, value))
}

fn parse_import_variables(input: &str) -> IResult<&str, Vec<Ast>> {
    let (input, values) = many0(terminated(
        map(
            separated_pair(
                ws(identifier),
                ws(tag(":")),
                alt((ws(address), ws(hex_literal), ws(bracketed_value))),
            ),
            |(name, value)| Ast::ImportVar {
                name,
                value: Box::new(value),
            },
        ),
        tuple((tag(","), multispace0)),
    ))(input)?;

    Ok((input, values))
}

// TODO: this requires a trailing comma, it shouldn't
pub fn parse_import(input: &str) -> IResult<&str, Ast> {
    let (input, _) = multispace0(input)?;

    let (input, _) = tag("import")(input)?;
    let (input, _) = space0(input)?;

    let (input, path) = parse_path(input)?;
    let (input, _) = space0(input)?;

    let (input, name) = identifier(input)?;
    let (input, _) = space0(input)?;

    let (input, address) = address(input)?;
    let (input, _) = space0(input)?;

    let (input, _) = tag("{")(input)?;
    let (input, _) = multispace0(input)?;

    let (input, variables) = parse_import_variables(input)?;

    let (input, _) = tag("}")(input)?;
    let (input, _) = multispace0(input)?;

    Ok((
        input,
        Ast::Import {
            name,
            path,
            variables,
            address: Box::new(address),
        },
    ))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_path() {
        let simple_file = "\"file.aya\"";
        let (input, path) = parse_path(simple_file).unwrap();
        assert!(input.is_empty());
        assert_eq!(path, "file.aya");

        let complex_path = "\"./../some_path/some/./../file.aya\"";
        let (input, path) = parse_path(complex_path).unwrap();
        assert!(input.is_empty());
        assert_eq!(path, "./../some_path/some/./../file.aya");
    }

    #[test]
    fn test_parse_import() {
        let input = "import \"./path.aya\" module_name &fefe {}";
        let (input, import) = parse_import(input).unwrap();
        assert!(input.is_empty());
        insta::assert_debug_snapshot!(import);
    }

    #[test]
    fn test_parse_import_with_variables() {
        let input = r#"
            import "./path.aya" module_name &fefe {
                variable_a: $C0D3,
                variable_b: [!other_variable],
                varaible_c: &FEFE,
                variable_d: [other_module.variable],
            }
        "#;
        let (input, import) = parse_import(input).unwrap();
        assert!(input.is_empty());
        insta::assert_debug_snapshot!(import);
    }
}
