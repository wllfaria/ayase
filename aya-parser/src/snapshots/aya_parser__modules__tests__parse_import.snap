---
source: aya-parser/src/modules.rs
expression: import
---
Import {
    name: "module_name",
    path: "./path.aya",
    address: Address(
        "fefe",
    ),
    variables: [],
}
